implementation/sequicity/tsd_net.py

```diff
@@ -291,22 +393,37 @@ def get_sparse_selective_input(self, x_input_np):
         return result
 
     def forward(self, z_enc_out, u_enc_out, u_input_np, m_t_input, degree_input, last_hidden, z_input_np):
+        """
+        This is an implementation of the decoder part of CopyNet. We want to
+        replace this with a Transformer Decoder with a Copy Mechanism.
+
+        Does:
+            1) Transformation of input into sparse - TODO: Look into this, might be important
+            2) Attention, GRU, generation/copy score = CopyNet - No need to understand closely, should be replaced.
+        """
+        # TODO: Document what exactly are the inputs and outputs of this forward function?
         sparse_z_input = Variable(self.get_sparse_selective_input(z_input_np), requires_grad=False)
 
         m_embed = self.emb(m_t_input)
-        
-        # z_enc_out are stacked (along the time-step axis) outputs of the BSpan decoder
+
+        # z_enc_out are stacked (along the time-step axis) belief spans (computed in BSpanDecoder)
+        # TODO: Why should this be a BSpanDecoder output, when the variable states:
+        #   - z_ = belief span
+        #   - enc_out = encoder output
+        #   - Would a better name be z_dec_out, or is this incorrect understanding of the code?
         z_context = self.attn_z(last_hidden, z_enc_out, mask=True, stop_tok=[self.vocab.encode('EOS_Z2')],
                                 inp_seqs=z_input_np)
-        # u_enc_out are output from the encoder
+
+        # u_enc_out are "user encoder output" = user inputs encoded by the Encoder (SimpleDynamicEncoder)
         u_context = self.attn_u(last_hidden, u_enc_out, mask=True, stop_tok=[self.vocab.encode('EOS_M')],
                                 inp_seqs=u_input_np)
-                                
+
         # TODO what is degree_input? it is coming from the data
```

Comment:
`degree_input` is the number of hits of search in knowledge base, represented as one-hot vector. 
`degree_size` was 6 when I ran the code.
Examples: `[0,0,1,0,0,0]` indicating 1 hit,  `[0,0,0,0,0,1]` indicating >=5 hits.